File: workbench_chatGPT/python/benchmark_sql_odbc.py
#!/usr/bin/env python3

import os, time, json, struct, sys
import pyodbc

from azure.identity import DefaultAzureCredential, DeviceCodeCredential, AzureCliCredential

SERVER = os.getenv("AZURE_SQL_SERVER")
DATABASE = os.getenv("AZURE_DATABASE")
TENANT = os.getenv("AZURE_TENANT_ID")
SCOPE = "https://database.windows.net/.default"

QUERY = os.getenv("BENCH_QUERY", "SELECT TOP 1000000 * FROM sys.objects a CROSS JOIN sys.objects b")

def _wide_token(token: str) -> bytes:
    w = token.encode("utf-16-le")
    return struct.pack("=i", len(w)) + w

def get_token():
    # Try Default (excluding interactive browser), then CLI, then Device Code
    try:
        cred = DefaultAzureCredential(exclude_interactive_browser_credential=True)
        return cred.get_token(SCOPE).token
    except Exception:
        pass
    try:
        return AzureCliCredential().get_token(SCOPE).token
    except Exception:
        pass
    return DeviceCodeCredential(tenant_id=TENANT).get_token(SCOPE).token

def main():
    if not SERVER or not DATABASE or not TENANT:
        print("Missing env: AZURE_SQL_SERVER / AZURE_DATABASE / AZURE_TENANT_ID", file=sys.stderr)
        sys.exit(2)

    token = get_token()
    conn_str = (
        f"Driver={{ODBC Driver 18 for SQL Server}};"
        f"Server={SERVER};Database={DATABASE};Encrypt=yes;TrustServerCertificate=no;"
    )

    t0 = time.perf_counter()
    conn = pyodbc.connect(conn_str, attrs_before={1256: _wide_token(token)})
    t_connect = (time.perf_counter() - t0) * 1000.0

    cur = conn.cursor()
    t1 = time.perf_counter()
    rows = cur.execute(QUERY).fetchall()
    t_query = (time.perf_counter() - t1)

    # Minimal metrics
    result = {
        "engine": "azure-sql-odbc",
        "connect_ms": round(t_connect, 2),
        "query_s": round(t_query, 3),
        "rows_returned": len(rows),
        "query_sample": QUERY[:120] + ("..." if len(QUERY) > 120 else ""),
    }
    print(json.dumps(result, indent=2))

if __name__ == "__main__":
    main()

File: workbench_chatGPT/R/benchmark_sql_odbc.R
#!/usr/bin/env Rscript
suppressMessages({
  library(DBI); library(odbc); library(jsonlite)
})

server <- Sys.getenv("AZURE_SQL_SERVER")
database <- Sys.getenv("AZURE_DATABASE")
query <- Sys.getenv("BENCH_QUERY", "SELECT TOP 1000000 * FROM sys.objects a CROSS JOIN sys.objects b")

if (server == "" || database == "") {
  stop("Missing env: AZURE_SQL_SERVER / AZURE_DATABASE", call. = FALSE)
}

t0 <- proc.time()[["elapsed"]]
con <- dbConnect(odbc::odbc(),
                 Driver = "ODBC Driver 18 for SQL Server",
                 Server = server,
                 Database = database,
                 Encrypt = "yes",
                 TrustServerCertificate = "no",
                 Authentication = "ActiveDirectoryDeviceCode")
t_connect <- (proc.time()[["elapsed"]] - t0)

t1 <- proc.time()[["elapsed"]]
res <- dbGetQuery(con, query)
t_query <- (proc.time()[["elapsed"]] - t1)

out <- list(
  engine = "azure-sql-odbc-r",
  connect_s = round(t_connect, 3),
  query_s = round(t_query, 3),
  rows_returned = nrow(res),
  query_sample = paste0(substr(query, 1, 120), ifelse(nchar(query) > 120, "...", ""))
)
cat(prettify(toJSON(out, auto_unbox = TRUE)), "\n")
dbDisconnect(con)
